File: api/internal/handlers/secretkeys.go
package handlers

import (
	"fmt"
	"regexp"
	"sort"
)

// Kubernetes rejects Secret keys that are longer than 253 characters or that contain
// anything outside [-._a-zA-Z0-9]. That rejection happens in the operator, against the
// API server, long after the value was accepted here — leaving the VestaSecret stored
// but unsyncable and the reconciler retrying forever. Validate at the write path instead
// so a bad key is a 400 the caller sees immediately.
const maxSecretKeyLength = 253

var secretKeyPattern = regexp.MustCompile(`^[-._a-zA-Z0-9]+$`)

func validateSecretKey(key string) error {
	switch {
	case key == "":
		return fmt.Errorf("secret key must not be empty")
	case len(key) > maxSecretKeyLength:
		return fmt.Errorf("secret key %s is %d characters long; keys must be no more than %d "+
			"(a key this long usually means a value was entered in the key field)",
			describeSecretKey(key), len(key), maxSecretKeyLength)
	case key == "." || key == "..":
		return fmt.Errorf("secret key must not be %q", key)
	case !secretKeyPattern.MatchString(key):
		return fmt.Errorf("secret key %s is invalid: keys may contain only letters, digits, '-', '_' and '.'",
			describeSecretKey(key))
	}
	return nil
}

// validateSecretKeys reports every offending key at once, so importing a .env file does
// not turn into one round trip per mistake.
func validateSecretKeys(keys []string) error {
	var bad []string
	for _, k := range keys {
		if err := validateSecretKey(k); err != nil {
			bad = append(bad, err.Error())
		}
	}
	if len(bad) == 0 {
		return nil
	}
	sort.Strings(bad)
	if len(bad) == 1 {
		return fmt.Errorf("%s", bad[0])
	}
	msg := fmt.Sprintf("%d invalid secret keys:", len(bad))
	for _, b := range bad {
		msg += "\n  - " + b
	}
	return fmt.Errorf("%s", msg)
}

func validateSecretDataKeys(data map[string]string) error {
	keys := make([]string, 0, len(data))
	for k := range data {
		keys = append(keys, k)
	}
	return validateSecretKeys(keys)
}

func validateSecretDataKeysAny(data map[string]interface{}) error {
	keys := make([]string, 0, len(data))
	for k := range data {
		keys = append(keys, k)
	}
	return validateSecretKeys(keys)
}

// describeSecretKey truncates before echoing a key back. An over-long key is typically a
// pasted secret value, and the error message travels into API responses and logs.
func describeSecretKey(key string) string {
	const shown = 24
	runes := []rune(key)
	if len(runes) <= shown {
		return fmt.Sprintf("%q", key)
	}
	return fmt.Sprintf("%q (truncated)", string(runes[:shown])+"…")
}

File: ui/src/lib/secretKeys.ts
// Kubernetes rejects Secret keys longer than 253 characters or containing anything
// outside [-._a-zA-Z0-9]. That rejection happens in the operator, far from where the key
// was typed, and leaves a VestaSecret that can never sync. Validate at entry instead.
export const MAX_SECRET_KEY_LENGTH = 253

const SECRET_KEY_PATTERN = /^[-._a-zA-Z0-9]+$/

/** Returns an error message for an invalid key, or null when the key is usable. */
export function secretKeyError(key: string): string | null {
  if (!key) return 'Key must not be empty'
  if (key.length > MAX_SECRET_KEY_LENGTH) {
    // A key this long is nearly always a value pasted into the key field, so say so
    // rather than just quoting the limit.
    return `Key is ${key.length} characters (max ${MAX_SECRET_KEY_LENGTH}) — did you paste a value into the key field?`
  }
  if (key === '.' || key === '..') return `Key must not be "${key}"`
  if (!SECRET_KEY_PATTERN.test(key)) {
    return "Key may contain only letters, digits, '-', '_' and '.'"
  }
  return null
}

export function isValidSecretKey(key: string): boolean {
  return secretKeyError(key) === null
}

/**
 * Parses .env-style text into key/value pairs, separating out lines whose key Kubernetes
 * would reject. A PEM or other multi-line value pasted unquoted lands here: its base64
 * continuation lines end in "=" padding, so they parse as a key with an empty value.
 */
export function parseEnvContent(content: string): {
  data: Record<string, string>
  invalidKeys: string[]
} {
  const data: Record<string, string> = {}
  const invalidKeys: string[] = []

  for (const line of content.split('\n')) {
    const trimmed = line.trim()
    if (!trimmed || trimmed.startsWith('#')) continue
    const eqIdx = trimmed.indexOf('=')
    if (eqIdx === -1) continue
    const key = trimmed.slice(0, eqIdx).trim()
    let value = trimmed.slice(eqIdx + 1).trim()
    if (value.startsWith('"') && value.endsWith('"') && value.length >= 2) {
      // Double quotes are the escaping form formatEnvContent emits, so undo the escapes
      // here. Without this the two functions disagree and copied output re-imports wrong.
      value = unescapeEnvValue(value.slice(1, -1))
    } else if (value.startsWith("'") && value.endsWith("'") && value.length >= 2) {
      value = value.slice(1, -1)
    }
    if (!key) continue
    if (!isValidSecretKey(key)) {
      invalidKeys.push(key)
      continue
    }
    data[key] = value
  }

  return { data, invalidKeys }
}

/**
 * A value needing quotes to survive a bare KEY=value line: parseEnvContent trims each
 * side, treats a leading # as a comment, and splits on newlines, so any of those
 * characters would come back different from what went out.
 */
const NEEDS_QUOTING = /^\s|\s$|[\n\r\t"'#\\]/

function escapeEnvValue(value: string): string {
  return value
    .replace(/\\/g, '\\\\')
    .replace(/"/g, '\\"')
    .replace(/\n/g, '\\n')
    .replace(/\r/g, '\\r')
    .replace(/\t/g, '\\t')
}

function unescapeEnvValue(value: string): string {
  return value.replace(/\\([nrt"'\\])/g, (_, ch) => {
    switch (ch) {
      case 'n': return '\n'
      case 'r': return '\r'
      case 't': return '\t'
      default: return ch
    }
  })
}

/**
 * Serialises key/value pairs to .env text — the inverse of parseEnvContent, so copied
 * output pastes straight back into an import box. Keys are sorted to make two copies of
 * the same config diffable; Kubernetes does not preserve map order anyway.
 */
export function formatEnvContent(data: Record<string, string>): string {
  return Object.keys(data)
    .sort()
    .map((key) => {
      const value = data[key] ?? ''
      return `${key}=${NEEDS_QUOTING.test(value) ? `"${escapeEnvValue(value)}"` : value}`
    })
    .join('\n')
}

/** Truncates a key for display. An over-long key is usually secret material. */
export function truncateSecretKey(key: string, shown = 24): string {
  return key.length <= shown ? key : `${key.slice(0, shown)}… (${key.length} chars)`
}
